adapter/authzOpa/authzOpa.go

```diff
@@ -0,0 +1,183 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package authzOpa // import "istio.io/mixer/adapter/authzOpa"
+
+// NOTE: This adapter will eventually be auto-generated so that it automatically supports all CHECK and QUOTA
+//       templates known to Mixer. For now, it's manually curated.
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	rpc "github.com/googleapis/googleapis/google/rpc"
+	"github.com/open-policy-agent/opa/ast"
+	"github.com/open-policy-agent/opa/rego"
+
+	"istio.io/mixer/adapter/authzOpa/config"
+	"istio.io/mixer/pkg/adapter"
+	"istio.io/mixer/template/authz"
+)
+
+type (
+	builder struct {
+		adapterConfig *config.Params
+	}
+
+	handler struct {
+		policy      string
+		checkMethod string
+
+		context  context.Context
+		compiler *ast.Compiler
+
+		env adapter.Env
+	}
+)
+
+///////////////// Configuration Methods ///////////////
+
+func (b *builder) SetAuthzTypes(map[string]*authz.Type) {}
+
+func (b *builder) SetAdapterConfig(cfg adapter.Config) {
+	b.adapterConfig = cfg.(*config.Params)
+}
+
+func (b *builder) Validate() (ce *adapter.ConfigErrors) {
+	if len(b.adapterConfig.CheckMethod) == 0 {
+		ce = ce.Appendf(GetInfo().Name,
+			"CheckMethod was not configured")
+	}
+
+	parsed, err := ast.ParseModule("", string(b.adapterConfig.Policy))
+	if err != nil {
+		ce = ce.Appendf(GetInfo().Name,
+			"Failed to parse the OPA policy: %v", err)
+	}
+
+	compiler := ast.NewCompiler()
+	compiler.Compile(map[string]*ast.Module{"": parsed})
+	if compiler.Failed() {
+		ce = ce.Appendf(GetInfo().Name,
+			"Failed to compile the OPA policy: %v", compiler.Errors)
+	}
+
+	return nil
+}
+
+func (b *builder) Build(context context.Context, env adapter.Env) (adapter.Handler, error) {
+	h := handler{
+		policy:      b.adapterConfig.Policy,
+		checkMethod: b.adapterConfig.CheckMethod,
+		env:         env,
+	}
+
+	parsed, err := ast.ParseModule("", string(h.policy))
+	if err != nil {
+		return nil, errors.New(fmt.Sprintf(GetInfo().Name,
+			"Failed to parse the OPA policy: %v", err))
+	}
+
+	compiler := ast.NewCompiler()
+	compiler.Compile(map[string]*ast.Module{"": parsed})
+	if compiler.Failed() {
+		return nil, errors.New(fmt.Sprintf(GetInfo().Name,
+			"Failed to compile the OPA policy: %v", compiler.Errors))
+	}
+
+	h.compiler = compiler
+	h.context = context
+
+	return h, nil
+}
+
+////////////////// Runtime Methods //////////////////////////
+
+func (h handler) HandleAuthz(context context.Context, instance *authz.Instance) (adapter.CheckResult, error) {
+	variables := make(map[string]interface{})
```

Comment:
You should probably assert in your `Validate` func that no key names for subjects, resources, or verbs overlap. Otherwise you'll lose data at runtime. The names for all of these are available in the `Types` handed to you `SetAuthzTypes`.

Reply:
Where I can validate configurations for authz template? authzOpa adapter gets configuration it's own from the SetAdapterConfig method.

Reply:
The flow to initialize a new adapter, validate config is this:

1. `Set_Types` is called for each template you support (so in this case, just `SetAuthzTypes`). The types you're provided describe the shape (including the label types and names) of every object your adapter will ever see at runtime (when those types change we close the current instance and create a new one).
2. `SetAdapterConfig` is called
3. `Validate` is called; if it returns a non-nil error the config update is rejected and reported as bad.
4. `Build` is called; if if returns a non-nil error then we continue serving with the previous instance of the adapter (effectively rejecting the new config).

So if you just tuck the `authz.Type`s you get during `SetAuthzTypes` into a field on the builder like you do for your adapter config, you'll have both in hand during `Validate` to assert that no keys overlap.